docs/proposals/deletions_object_storage.md

```diff
@@ -0,0 +1,86 @@
+2. [https://github.com/thanos-io/thanos/issues/903](https://github.com/thanos-io/thanos/issues/903)
+
+# Summary
+
+This design document proposes deletion of series for object storage in Thanos. This feature mainly causes changes in the Compactor and Store components and the object storage itself where the changes are expected to be reflected.
+
+
+# Motivation
+
+The main motivation for considering deletions in the object storage are the following use cases
+
+
+
+*   **Accidental insertion of confidential data:** This is a scenario where a user accidentally inserts confidential data and wants to delete it immediately. In this case, the user expects their request to accelerate an immediate deletion of the series pertaining to the blocks concerned with that specific data to be deleted.
+*   **GDPR:** Masking data and eventual deletion is expected.
+*   **Deletions to sustain user requirements:** Let’s assume the user has some data which leads to some unexpected results or causes performance degradation (due to high cardinality) and the user wants to restore the previous data set-up for obtaining the desired results. In this scenario, the user would want to send a request to mask the data for the time being as there isn’t a high priority requirement to delete the data but eventually during the compaction the user can expect the data to be deleted by the compactor not leading to any major performance issues.
+*   **Helps achieving series based retention (e.g rule aggregation) [#903](https://github.com/thanos-io/thanos/issues/903).**
+
+
+## Goals
+
+
+
+*   Unblock users and allow series deletion in the object storage using tombstones.
+*   Allowing undeletes for a default time duration and extending it as per the user requirements.
+*   Dealing with already downsampled and to be downsampled blocks with tombstones.
+*   Performing all the above operations at admin level.
+
+
+# Proposed Approach
+
+
+
+*   We propose to implement deletions using the tombstones approach.
+*   A user is expected to enter the following details for performing deletions:
+    *   **external labels**
+    *   **start timestamp**
+    *   **end timestamp** (start and end timestamps of the series data the user expects to be deleted)
+    *   **maximum waiting duration for performing deletions** where a default value is considered if explicitly not specified by the user. (only after this time passes the deletions are performed by the compactor in the next compaction cycle)
+*   The details are entered via a deletions API (good to have a web UI interface) and they are processed by the compactor to create a tombstone file, if there’s a match for the details entered. Afterwards the tombstone file is uploaded to object storage making it accessible to other components.
+*   If the data with the requested details couldn’t be found in the storage an error message is returned back as a response to the request.
```

Comment:
> If the data with the requested details couldn’t be found in the storage an error message is returned back as a response to the request.

This means the deletion API request is synchronous. The problem is that on large setups, processing this request may take much time and the request could timeout before done. I think an async approach would be more robust:

1. Deletion API returns a request ID
2. A second API allows to GET the status of a request ID

Reply:
We thought that the creation of tombstones is synchronous and should be able to respond fairly quickly. The deletion itself of times series is still asynchronous and will eventually happen.
The API has an option to list all tombstones and through that get their status.

Because deletion is asynchronous and the creation of tombstones synchronous but fairly quick, we don't think the overhead of a asynchronous approach for the tombstones creation is justified.